src/core/qgsmaplayerregistry.cpp

```diff
@@ -131,21 +131,23 @@ void QgsMapLayerRegistry::removeMapLayers( const QList<QgsMapLayer*>& layers )
     return;
 
   QStringList layerIds;
+  QList<QgsMapLayer*> layerList;
 
   Q_FOREACH ( QgsMapLayer* layer, layers )
   {
-    if ( layer )
+    // check layer and the registry contains it
+    if ( layer && mMapLayers.contains( layer->id() ) )
+    {
       layerIds << layer->id();
+      layerList << layer;
+    }
   }
 
   emit layersWillBeRemoved( layerIds );
-  emit layersWillBeRemoved( layers );
+  emit layersWillBeRemoved( layerList );
```

Comment:
There is still only one signal defined in the header

    void layersWillBeRemoved( const QStringList& theLayerIds );